### PR TITLE
Add Enterprise SONIC LLDP Interface module

### PR DESCRIPTION
**Pull request by m-nazeer**
##### SUMMARY
LLDP_INTERFACES Module implementation

##### GitHub Issues
N/A

##### ISSUE TYPE
Feature Pull Request

##### COMPONENT NAME
sonic_lldp_interfaces

##### OUTPUT

Regression Test HTML report: 
[lldp_interface_log.zip](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/15388789/lldp_interface_log.zip)

##### Checklist:

- [x] I have performed a self-review of my own code to ensure there are no formatting, linting, or security issues
- [x] I have verified that new and existing unit tests pass locally with my changes
- [x] I have not allowed coverage numbers to degenerate
- [x] I have maintained at least 90% code coverage
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have maintained backward compatibility or have provided any relevant "breaking_changes" descriptions in a "fragment" file in the "changelogs/fragments" directory of this repository.
- [x] I have provided a summary for this PR in valid "fragment" file format in the "changelogs/fragments" directory of this repository branch. Reference : [Ansible Change Log Document](https://docs.ansible.com/ansible/devel/community/development_process.html#changelogs-how-to)

### CODE REVIEW

**Comment by kerry-meyer** [comment]
Closing and re-opening to clear "failure" for a missing fragment file.

**Comment by kerry-meyer** [comment]
To proceed with the code review for this resource module, please push changes for the following items:

- Re-sync your fork and branch to pull in updates from the "main" branch that are the probable cause of the flagged sanity failures for python3.10, 3.11, and 3.12 with the ansible-core "devel" version.
- Add Unit Test coverage. Use the UT implementation for the DHCP Relay module as an example of how to do this.

While you are working on the UT changes, we may go ahead with checking the changes that are in the current change set and  posting an initial set of comments/questions (if any) for the current code. The PR can not be approved or merged, however, until the items described above have been completed.

File: plugins/module_utils/network/sonic/facts/lldp_interfaces/lldp_interfaces.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+                            lldp_interface_data['med_tlv_select']['power_management'] = False
+                        if 'openconfig-lldp-ext:MDI_POWER' in config['openconfig-lldp-ext:suppress-tlv-advertisement']:
+                            lldp_interface_data['tlv_select']['power_management'] = False
+                    if 'openconfig-lldp-ext:management-address-ipv4' in config:
+                        lldp_tlv_dict['ipv4_management_address'] = config.get('openconfig-lldp-ext:management-address-ipv4')
+                    if 'openconfig-lldp-ext:management-address-ipv6' in config:
+                        lldp_tlv_dict['ipv6_management_address'] = config.get('openconfig-lldp-ext:management-address-ipv6')
+                    if any(lldp_tlv_dict.values()):
+                        lldp_interface_data['tlv_set'] = lldp_tlv_dict
+                    lldp_interface_configs.append(lldp_interface_data.copy())
```
Comment (mingjunzhang2019) [inline plugins/module_utils/network/sonic/facts/lldp_interfaces/lldp_interfaces.py:125]: the "copy" may not be needed.

File: plugins/module_utils/network/sonic/config/lldp_interfaces/lldp_interfaces.py [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+        """
+        commands = []
+        requests = []
+        if not diff:
+            return commands, requests
+        commands = want
+        for command in commands:
+            # Get interface name and make all configs to default in that interface
+            command = {"name": command["name"]}
+            requests.extend(self.get_delete_specific_lldp_interfaces_param_requests(command, have))
```
Comment (mingjunzhang2019) [inline plugins/module_utils/network/sonic/config/lldp_interfaces/lldp_interfaces.py:197]: the "commands" should be updated.

File: plugins/module_utils/network/sonic/config/lldp_interfaces/lldp_interfaces.py [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+    def _state_replaced(self, diff, want, have):
+        """ The command generator when state is replaced
+        :rtype: A list
+        :returns: the commands necessary to migrate the current configuration
+                  to the desired configuration
+        """
+        commands = []
+        requests = []
+        if not diff:
+            return commands, requests
```
Comment (mingjunzhang2019) [inline plugins/module_utils/network/sonic/config/lldp_interfaces/lldp_interfaces.py:192]: diff = get_diff(want, have). If want is subset of have, some configs should be replaced.

File: plugins/module_utils/network/sonic/config/lldp_interfaces/lldp_interfaces.py [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+        """ The command generator when state is overridden
+
+        :rtype: A list
+        :returns: the commands necessary to migrate the current configuration
+                  to the desired configuration
+        """
+        commands = []
+        requests = []
+        if not diff:
+            return commands, requests
```
Comment (mingjunzhang2019) [inline plugins/module_utils/network/sonic/config/lldp_interfaces/lldp_interfaces.py:216]: diff = get_diff(want, have), if want is subset of have, the configs should be overridden.

File: plugins/module_utils/network/sonic/config/lldp_interfaces/lldp_interfaces.py [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+        configurations in the chassis
+        """
+        default_dict = {'tlv_select': {'power_management': True}, 'med_tlv_select': {'network_policy': True, 'power_management': True}, 'enable': True}
+        requests = []
+        conf = copy.deepcopy(have)
+        for cfg in conf:
+            del cfg['name']
+            if default_dict != cfg:
+                return [{'path': self.lldp_interfaces_path, 'method': DELETE}]
+        return requests
```
Comment (mingjunzhang2019) [inline plugins/module_utils/network/sonic/config/lldp_interfaces/lldp_interfaces.py:240]: The login shows either one "cfg" in deleting  request or request is []. Please check.

**Review by kerry-meyer — CHANGES_REQUESTED** [review]
I'm changing the state on this PR to "Request changes" to make it easier to see from the top level PR view.

Please address the comments posted by Mingjun Zhang either with the corresponding modifications in the change set or with response comments (for cases in which you would like clarification of the posted comment or do not agree with it).

File: plugins/module_utils/network/sonic/config/lldp_interfaces/lldp_interfaces.py (hunk 2 above)
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/config/lldp_interfaces/lldp_interfaces.py]: The rewritten version of this block looks fine.

Resolved.

File: plugins/module_utils/network/sonic/config/lldp_interfaces/lldp_interfaces.py [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+        :returns: the commands necessary to migrate the current configuration
+                  to the desired configuration
+        """
+        commands = []
+        requests = []
+        if not diff:
+            return commands, requests
+        commands = have
+        # Default all existing interface configuration
+        requests.extend(self.get_delete_lldp_interfaces_complete_requests(commands))
```
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/config/lldp_interfaces/lldp_interfaces.py]: The rewritten version of this block looks fine.

Resolved.

File: plugins/module_utils/network/sonic/config/lldp_interfaces/lldp_interfaces.py (hunk 3 above)
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/config/lldp_interfaces/lldp_interfaces.py]: The rewritten version of this block looks fine.

Resolved.

File: plugins/module_utils/network/sonic/config/lldp_interfaces/lldp_interfaces.py (hunk 4 above)
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/config/lldp_interfaces/lldp_interfaces.py]: Although the revised block of code looks correct, the posted test results do not.

Please post test results for the current version of the change set.

File: plugins/module_utils/network/sonic/config/lldp_interfaces/lldp_interfaces.py (hunk 5 above)
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/config/lldp_interfaces/lldp_interfaces.py:266]: I see that the logic is intended to delete all LLDP interface configuration with one REST request if at least one interface is found with non-default configuration. This looks okay to me.

There is, however, currently no test case for this. Please add one.

File: plugins/module_utils/network/sonic/facts/lldp_interfaces/lldp_interfaces.py (hunk 1 above)
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/facts/lldp_interfaces/lldp_interfaces.py:125]: Please address this comment.

**Review by kerry-meyer — CHANGES_REQUESTED** [review]
Thank you for making incremental changes to address Mingjun's review comments.

I have marked some of the "Issues" posted by Mingjun as "Resolved", but some are not, and I have posted a couple of additional comments.

- Please address the remaining open issues and post updated test results. In the posted test results, please format the summary output in a way that displays the 'before' and 'after' output without truncation. (Landscape, scaled as needed.)
- Please ensure that overridden/replaced test cases are causing deletion of some attributes within existing interface configurations. (This may already be the case, but it's not easy to see from the currently posted test results or by 'walking' the test list.)
- Please fix the UT failures and sanity failures.
- Please re-sync the PR branch to pick up changes from the 'main' branch and to resolve any conflicts that currently exist for shared files.

File: plugins/module_utils/network/sonic/config/lldp_interfaces/lldp_interfaces.py (hunk 5 above)
Comment (m-nazeer) [reply plugins/module_utils/network/sonic/config/lldp_interfaces/lldp_interfaces.py:267]: Added a new case for this scenario

File: plugins/module_utils/network/sonic/config/lldp_interfaces/lldp_interfaces.py (hunk 4 above)
Comment (m-nazeer) [reply plugins/module_utils/network/sonic/config/lldp_interfaces/lldp_interfaces.py]: Attached the latest result

File: plugins/module_utils/network/sonic/facts/lldp_interfaces/lldp_interfaces.py (hunk 1 above)
Comment (m-nazeer) [reply plugins/module_utils/network/sonic/facts/lldp_interfaces/lldp_interfaces.py]: Addressed the review comment

Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/facts/lldp_interfaces/lldp_interfaces.py]: Verified.

Thank you.

File: plugins/module_utils/network/sonic/config/lldp_interfaces/lldp_interfaces.py (hunk 4 above)
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/config/lldp_interfaces/lldp_interfaces.py]: The current code and posted test results look correct for this case now.

Resolved.

File: plugins/module_utils/network/sonic/config/lldp_interfaces/lldp_interfaces.py (hunk 5 above)
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/config/lldp_interfaces/lldp_interfaces.py:267]: The posted test result looks correct.

Resolved.

**Review by kerry-meyer — APPROVED** [review]
The code for the current proposed change set, the test case coverage, and all test results look good.

Approved.

Thank you for providing the LLDP Interfaces resource module for the Dell enterprise_sonic resource module collection.

**Pull request merged**